### PR TITLE
Add public api doc for create_repository_using_definitions_args, update typehints, and cleanup Definitions docstring

### PR DESCRIPTION
**Pull request by schrockn**
### Summary & Motivation

User understably expected an API I referenced to have an entry in our docs site. See discussion in https://github.com/dagster-io/dagster/issues/12157.

Also I noticed that `create_repository_using_definitions_args` typehint was out of date and that `Definitions` docstring had some formatting issues

### How I Tested These Changes

Eyes

Preview: https://dagster-git-add-createrepositoryusingdefinition-838934-elementl.vercel.app/_apidocs/definitions

### CODE REVIEW

**Comment by schrockn** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12176** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12176" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/12176?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/definitions_class.py [hunk 1]
```diff
@@ -36,17 +36,34 @@ def create_repository_using_definitions_args(
...
     resources: Optional[Mapping[str, Any]] = None,
-    executor: Optional[ExecutorDefinition] = None,
+    executor: Optional[Union[ExecutorDefinition, Executor]] = None,
     loggers: Optional[Mapping[str, LoggerDefinition]] = None,
 ) -> Union[RepositoryDefinition, PendingRepositoryDefinition]:
     """
-    For users who, for the time being, want to continue to use multiple named repositories in
-    a single code location, you can use this function. The behavior (e.g. applying resources to
-    all assets) are identical to :py:class:`Definitions` but this returns a named repository.
+    Create a named repository using the same arguments as :py:class:`Definitions`. This enables
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/definitions_class.py:49]: I imagine that a new user who comes across this and has never heard of repositories might think this is confusing and that they might be missing something in their understanding of Dagster. What do you think about including some text to clarify that this refers to a legacy concept?

**Review by sryza — APPROVED** [review]
Looks great

File: python_modules/dagster/dagster/_core/definitions/definitions_class.py [hunk 2]
```diff
@@ -36,17 +36,39 @@ def create_repository_using_definitions_args(
...
     loggers: Optional[Mapping[str, LoggerDefinition]] = None,
 ) -> Union[RepositoryDefinition, PendingRepositoryDefinition]:
     """
-    For users who, for the time being, want to continue to use multiple named repositories in
-    a single code location, you can use this function. The behavior (e.g. applying resources to
-    all assets) are identical to :py:class:`Definitions` but this returns a named repository.
+    Create a named repository using the same arguments as :py:class:`Definitions`. In older
+    versions of Dagster, repositories were the mechanism for organizing assets, schedules, sensors,
+    and jobs. There could be many repositories per code location. This was a complicated ontology but
+    gave users a way to organize code location that contained large number of heterogenous definitions.
```
Comment (erinkcochran87) [inline python_modules/dagster/dagster/_core/definitions/definitions_class.py:52]:
```suggestion
    gave users a way to organize code locations that contained large numbers of heterogenous definitions.
```

A few small typos

**Review by erinkcochran87 — APPROVED** [review]
One comment about a few small typos, but otherwise lgtm!

**Pull request merged**